[ILM] Add documentation for error handling in ILM (#35426)
elastic/elasticsearch: docs/reference/ilm/error-handling.asciidoc

[role="xpack"]
[[index-lifecycle-error-handling]]
== Index Lifecycle Error Handling

During Index Lifecycle Management's execution of the policy for an index, it's
possible for a step to encounter an error during its execution. When this
happens, ILM will move the management state into an "error" step. This halts
further execution of the policy and gives an administrator the chance to address
any issues with the policy, index, or cluster.

An example will be helpful in illustrating this, imagine the following policy
has been created by a user:

[source,js]
--------------------------------------------------
PUT _ilm/policy/shrink-the-index
{
  "policy": {
    "phases": {
      "warm": {

>         "min_age": "5m",

Reviewer: I think we should use values of `min_age` that we expect users to use in documentation. Can we use a value like 5 days here?

Author: Sure, no problem

        "actions": {
          "shrink": {
            "number_of_shards": 4
          }
        }
      }
    }
  }
}
--------------------------------------------------
// CONSOLE
// TEST

This policy waits until the index is at least 5 minutes old, and then shrinks
the index to 4 shards.

Now imagine that a user creates a new index "myindex" with two primary shards,
telling it to use the policy they have created:

[source,js]
--------------------------------------------------
PUT /myindex
{
  "settings": {
    "index.number_of_shards": 2,
    "index.lifecycle.name": "shrink-the-index"
  }
}
--------------------------------------------------
// CONSOLE
// TEST[continued]

After five minutes has passed, ILM will attempt to shrink this index from 2

> shards to 4, which is an invalid amount. When this occurs, ILM will move this

Reviewer: ```suggestion
shards to 4, which is invalid since the shrink action cannot increase the
number of shards. When this occurs, ILM will move this
```

has been created by a user:

[source,js]
--------------------------------------------------
PUT _ilm/policy/shrink-the-index
{
  "policy": {
    "phases": {
      "warm": {
        "min_age": "5d",
        "actions": {
          "shrink": {
            "number_of_shards": 4
          }
        }
      }
    }
  }
}
--------------------------------------------------
// CONSOLE
// TEST

This policy waits until the index is at least 5 days old, and then shrinks
the index to 4 shards.

Now imagine that a user creates a new index "myindex" with two primary shards,
telling it to use the policy they have created:

[source,js]
--------------------------------------------------
PUT /myindex
{
  "settings": {
    "index.number_of_shards": 2,
    "index.lifecycle.name": "shrink-the-index"
  }
}
--------------------------------------------------
// CONSOLE
// TEST[continued]

After five days have passed, ILM will attempt to shrink this index from 2
shards to 4, which is invalid since the shrink action cannot increase the
number of shards. When this occurs, ILM will move this
index to the "error" step. Once an index is in this step, information about the

> reason for the error can be retrieved from the <<ilm-explain,ILM Explain API>>:

Reviewer 2: ```suggestion
reason for the error can be retrieved from the <<ilm-explain-lifecycle,ILM Explain API>>:
```

policy entails updating the existing policy to use one instead of four for
the targeted number of shards.

[source,js]
--------------------------------------------------
PUT _ilm/policy/shrink-the-index
{
  "policy": {
    "phases": {
      "warm": {
        "min_age": "5d",
        "actions": {
          "shrink": {
            "number_of_shards": 1
          }
        }
      }
    }
  }
}
--------------------------------------------------
// CONSOLE
// TEST[continued]

=== Retrying failed index lifecycle management steps

Once the underlying issue that caused an index to move to the error step has
been corrected, index lifecycle management must be told to retry the step to see
if it can progress further. This is accomplished by invoking the retry API

[source,js]
--------------------------------------------------
POST /myindex/_ilm/retry
--------------------------------------------------
// CONSOLE
// TEST[skip:we can't be sure the index is ready to be retried at this point]

Once this has been issue, index lifecycle management will asynchronously pick up
on the step that is in a failed state, attempting to re-run it. The

> <<ilm-explain,ILM Explain API>> can again be used to monitor the status of

Reviewer 2: ```suggestion
<<ilm-explain-lifecycle,ILM Explain API>> can again be used to monitor the status of
```
